stew/results.nim

```diff
@@ -953,25 +953,52 @@ template isNone*(o: Opt): bool =
 # Syntactic convenience
 
 template `?`*[T, E](self: Result[T, E]): auto =
+  mixin assignResult
   ## Early return - if self is an error, we will return from the current
   ## function, else we'll move on..
   ##
   ## ```
   ## let v = ? funcWithResult()
   ## echo v # prints value, not Result!
   ## ```
+  ##
+  ## `assignResult` extension point
+  ##
+  ## If a template / function named `assignResult` exists in the scope, it will
+  ## be called instead of assigning the result - this can be used to intercept
+  ## the assignment to `result` that implicitly happens on early return in case
+  ## of error.
+  ##
+  ## To prevent conflicts, it is recommended that `assignResult` is declared
+  ## close to the scope of `?` (as opposed to a globally visible symbol)
+  ##
+  ## ```nim
+  ## proc f(): Result[...] =
+  ##   template assignResult(v: Result) = ...
+  ##   let a = ? f2()
+  ## ```
+  ##
   ## Experimental
   # TODO the v copy is here to prevent multiple evaluations of self - could
   #      probably avoid it with some fancy macro magic..
   let v = (self)
   if not v.o:
-    when typeof(result) is typeof(v):
-      return v
-    else:
-      when E is void:
-        return err(typeof(result))
+    when declared(assignResult):
```

Comment:
I'm not sure this `declared` check would actually work for the Chronos use-case. Have you tested it?

Reply:
well, there's a test in https://github.com/status-im/nim-chronos/pull/297 - not sure about more complex cases, anything in particular you're worried about?

Reply:
Ah, OK, I didn't realize the Chronos PR was tested precisely with this version of `Result` (there was the possibility that `assignResult` existed a globally defined template that is overloaded in Chronos).

It would be good to test additional use cases such as generic async procs or async closures defined in generic functions as these trigger substantially different symbol lookup paths in the compiler (which may trip the `declared` check here).

Reply:
added a generic test which seems to work for now - needed an extra mixin _inside_ `assignResult` to not complain about the `result` symbol